### PR TITLE
Refactor Metta to be Single-Discrete in Python (because multi-discrete caused invalid actions and is ugly)

### PR DESCRIPTION
**Pull request by relh**
- MettaGrid core/C++ now builds a flattened action catalogue (build_flat_action_catalog), exposes a Discrete(n) action space, maps flat IDs back to (type,arg) when stepping, and records
    the flattened index in observation features. Python bindings (MettaGridCore) enforce (num_agents,) np.int32 actions and drop the old multi-discrete metadata.
  - Agents & RL stack were simplified to that contract: ActionProbs, transformer/sliding transformer, puffer policy, and the training loop all assume action_dim == 1, keep last-actions as
    scalar tensors, and remove padding/dtype gymnastics.
  - Cogames + tooling (player CLI, mettascope, demos, mock agents, tests) emit/consume scalar action IDs and use helper utilities like action_index(...). The extensive test suite under
    packages/mettagrid/tests was updated to use flat np.int32 arrays, and docs now describe the new semantics.

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1211568243608281)

### CODE REVIEW

File: agent/src/metta/agent/components/actor.py [hunk 1]
```diff
@@ -179,13 +175,6 @@ def forward_training(self, td: TensorDict, action: torch.Tensor) -> TensorDict:
...
-        """Convert (action_type, action_param) pairs to discrete indices."""
-        action_type_numbers = flattened_action[:, 0].long()
-        action_params = flattened_action[:, 1].long()
-        cumulative_sum = self.cum_action_max_params[action_type_numbers]
-        return cumulative_sum + action_type_numbers + action_params
-
-    def _convert_logit_index_to_action(self, logit_indices: torch.Tensor) -> torch.Tensor:
-        """Convert discrete logit indices back to (action_type, action_param) pairs."""
-        return self.action_index_tensor[logit_indices]
+    @property
```
Comment (rjwalters) [inline agent/src/metta/agent/components/actor.py:186]: do we need this helper?

File: docs/action_compatibility.md [hunk 2]
```diff
@@ -24,7 +24,7 @@ Actions in Metta consist of two components:
 
 ```
 1. Action Type Validation: 0 <= action_type < num_action_handlers
-2. Action Argument Validation: 0 <= action_arg <= max_action_args[action_type]
+2. Action Validation: ensure sampled indices are within `env.action_space.n`
```
Comment (rjwalters) [inline docs/action_compatibility.md]: maybe this line is just deleted?

File: metta/rl/training/core.py [hunk 3]
```diff
@@ -282,5 +289,5 @@ def add_last_action_to_td(self, td: TensorDict, env: TrainingEnvironment) -> Non
         if self.last_action is None or len(self.last_action) < env_ids.max() + 1:
             act_space = env.single_action_space
             act_dtype = torch.int32 if np.issubdtype(act_space.dtype, np.integer) else torch.float32
-            self.last_action = torch.zeros(env_ids.max() + 1, len(act_space.nvec), dtype=act_dtype, device=td.device)
+            self.last_action = torch.zeros(env_ids.max() + 1, 1, dtype=act_dtype, device=td.device)
```
Comment (rjwalters) [inline metta/rl/training/core.py]: should this be simplified to 1D?
`self.last_action = torch.zeros(env_ids.max() + 1, dtype=act_dtype, device=td.device)`

File: metta/sim/simulation.py [hunk 4]
```diff
@@ -248,30 +247,29 @@ def generate_actions(self) -> np.ndarray:
                 )
 
         with torch.no_grad():
-            policy_actions = self._get_actions_for_agents(self._policy_idxs.cpu(), self._policy)
+            policy_actions = self._get_actions_for_agents(self._policy_idxs.cpu(), self._policy).to(torch.long)
```
Comment (rjwalters) [inline metta/sim/simulation.py:250]: double check long vs int32?

File: agent/src/metta/agent/components/actor.py [hunk 5]
```diff
@@ -155,14 +145,20 @@ def forward_training(self, td: TensorDict, action: torch.Tensor) -> TensorDict:
...
                 td = td.reshape(td.batch_size.numel())
 
-        action_logit_index = self._convert_action_to_logit_index(action)
+        if action.dim() == 2 and action.size(1) == 1:
+            action = action.view(-1)
+
+        if action.dim() != 1:
+            raise ValueError(f"Expected flattened action indices, got shape {tuple(action.shape)}")
+
+        action_logit_index = action.to(dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (relh) [reply agent/src/metta/agent/components/actor.py:161]: pytorch indexing needs int64

File: metta/tools/play.py [hunk 6]
```diff
@@ -53,7 +53,7 @@ def invoke(self, args: dict[str, str]) -> int | None:
             initial_replay = sim.get_replay()
 
             current_step = 0
-            actions = np.zeros((env.num_agents, 2))
+            actions = np.zeros(env.num_agents, dtype=np.int64)
```
Comment (rjwalters) [inline metta/tools/play.py:56]: my recollection is that we must have int32 here... we have a type alias defined in mettagrid that might be useful

File: mettascope/server.py [hunk 7]
```diff
@@ -172,7 +172,7 @@ async def send_message(**kwargs):
 
         current_step = 0
         action_message = None
-        actions = np.zeros((env.num_agents, 2))
+        actions = np.zeros(env.num_agents, dtype=int)
```
Comment (rjwalters) [inline mettascope/server.py:175]: same issue with type

File: agent/src/metta/agent/components/actor.py [hunk 8]
```diff
@@ -114,22 +114,14 @@ class ActionProbs(nn.Module):
     def __init__(self, config: ActionProbsConfig):
         super().__init__()
         self.config = config
+        self.num_actions = 0
 
     def initialize_to_environment(
         self,
         env: Any,
```
[the remark this answers is not in the document]
Comment (relh) [reply agent/src/metta/agent/components/actor.py]: done! thanks 🙏🏼

File: packages/mettagrid/cpp/include/mettagrid/actions/attack.hpp [hunk 9]
```diff
@@ -30,90 +27,67 @@ struct AttackActionConfig : public ActionConfig {
 
 class Attack : public ActionHandler {
 public:
-  explicit Attack(const AttackActionConfig& cfg,
-                  const GameConfig* game_config,
-                  const std::string& action_name = "attack")
-      : ActionHandler(cfg, action_name), _defense_resources(cfg.defense_resources), _game_config(game_config) {
+  Attack(const AttackActionConfig& cfg,
```
Comment (rjwalters) [inline packages/mettagrid/cpp/include/mettagrid/actions/attack.hpp:30]: attack seems worse as a single discrete

**Review by rjwalters — COMMENTED** [review]
This pr is very large and seems to do more than strictly necessary. The complexity that is removed in policies feels like it is added back in mettagrid... overall I'm not convince this is an essential change

File: metta/sim/simulation.py [hunk 10]
```diff
@@ -29,6 +29,8 @@
 from metta.sim.utils import get_or_create_policy_ids
 from mettagrid import MettaGridEnv, dtype_actions
 
+TORCH_ACTION_DTYPE = torch.int32
```
Comment (rjwalters) [inline metta/sim/simulation.py]: I still see some int64 and long in files above...

File: packages/cogames/src/cogames/policy/lstm.py [hunk 11]
```diff
@@ -211,12 +209,10 @@ def step_with_state(
                 new_state = LSTMState.from_tuple(tuple_state, layers)
 
             # Sample action from the logits
-            actions: list[int] = []
-            for logit in logits:
-                dist = torch.distributions.Categorical(logits=logit)
-                actions.append(dist.sample().item())
+            dist = torch.distributions.Categorical(logits=logits)
+            action = np.int32(dist.sample().cpu().item())
```
Comment (rjwalters) [inline packages/cogames/src/cogames/policy/lstm.py:213]:
```suggestion
            action = dtype_actions(dist.sample().cpu().item())
```

File: packages/cogames/src/cogames/policy/random.py [hunk 12]
```diff
@@ -24,7 +24,7 @@ def step(self, obs: MettaGridObservation) -> MettaGridAction:
             A random action sampled from the action space
         """
         sample = self._action_space.sample()
-        return np.asarray(sample, dtype=np.int32)
+        return np.int32(sample)
```
Comment (rjwalters) [inline packages/cogames/src/cogames/policy/random.py:27]:
```suggestion
        return dtype_actions(sample)
```

File: packages/cogames/src/cogames/policy/token.py [hunk 13]
```diff
@@ -98,43 +98,38 @@ def forward_eval(
...
-
-            actions: list[int] = []
-            for logit in logits:
-                dist = torch.distributions.Categorical(logits=logit)
-                actions.append(dist.sample().item())
-
-            return np.array(actions, dtype=np.int32)
+            dist = torch.distributions.Categorical(logits=logits)
+            action = dist.sample().item()
+            return np.array(action, dtype=np.int32)
```
Comment (rjwalters) [inline packages/cogames/src/cogames/policy/token.py:132]: dtype_actions

**Review by rjwalters — APPROVED** [review]
I think this is a reasonable approach - a few nits on type unification but consider these optional
